<style>
:root {
  --bg:       #0d1b2a;
  --bg2:      #1a2a3a;
  --bg3:      #0a1628;
  --accent:   #1b91ff;
  --orange:   #e07b39;
  --green:    #2fca72;
  --gold:     #f0c040;
  --text:     #c8d8e8;
  --text2:    #8ab4d4;
  --border:   #1b3a5c;
  --code-bg:  #1e2d3d;
  --code-fg:  #7fdbca;
  --red:      #e05555;
}
body, .jp-RenderedHTMLCommon {
  background: var(--bg) !important;
  color: var(--text) !important;
  font-family: 'Segoe UI', Arial, sans-serif !important;
}
h1,h2,h3,h4 { color: #ffffff !important; }
h1 { border-bottom: 2px solid var(--accent) !important; padding-bottom: 8px !important; }
h2 { color: var(--accent) !important; }
h3 { color: var(--text2) !important; }
a  { color: var(--accent) !important; }
code {
  background: var(--code-bg) !important;
  color: var(--code-fg) !important;
  padding: 2px 6px !important;
  border-radius: 4px !important;
}
pre code { padding: 0 !important; }
pre {
  background: var(--code-bg) !important;
  border-left: 3px solid var(--accent) !important;
  border-radius: 6px !important;
  padding: 14px 18px !important;
}
table { border-collapse: collapse !important; width: 100% !important; }
th {
  background: var(--bg3) !important;
  color: var(--accent) !important;
  border: 1px solid var(--border) !important;
  padding: 8px 12px !important;
}
td {
  border: 1px solid var(--border) !important;
  padding: 8px 12px !important;
  color: var(--text) !important;
}
tr:nth-child(even) td { background: var(--bg2) !important; }
blockquote {
  border-left: 4px solid var(--gold) !important;
  background: #1a2a1a !important;
  padding: 10px 16px !important;
  border-radius: 0 8px 8px 0 !important;
  color: var(--text) !important;
}
</style>

<style>
:root {
  --bg:       #0d1b2a;
  --bg2:      #1a2a3a;
  --bg3:      #0a1628;
  --accent:   #1b91ff;
  --orange:   #e07b39;
  --green:    #2fca72;
  --gold:     #f0c040;
  --text:     #c8d8e8;
  --text2:    #8ab4d4;
  --border:   #1b3a5c;
  --code-bg:  #1e2d3d;
  --code-fg:  #7fdbca;
  --red:      #e05555;
}
body, .jp-RenderedHTMLCommon {
  background: var(--bg) !important;
  color: var(--text) !important;
  font-family: 'Segoe UI', Arial, sans-serif !important;
}
h1,h2,h3,h4 { color: #ffffff !important; }
h1 { border-bottom: 2px solid var(--accent) !important; padding-bottom: 8px !important; }
h2 { color: var(--accent) !important; }
h3 { color: var(--text2) !important; }
a  { color: var(--accent) !important; }
code {
  background: var(--code-bg) !important;
  color: var(--code-fg) !important;
  padding: 2px 6px !important;
  border-radius: 4px !important;
}
pre code { padding: 0 !important; }
pre {
  background: var(--code-bg) !important;
  border-left: 3px solid var(--accent) !important;
  border-radius: 6px !important;
  padding: 14px 18px !important;
}
table { border-collapse: collapse !important; width: 100% !important; }
th {
  background: var(--bg3) !important;
  color: var(--accent) !important;
  border: 1px solid var(--border) !important;
  padding: 8px 12px !important;
}
td {
  border: 1px solid var(--border) !important;
  padding: 8px 12px !important;
  color: var(--text) !important;
}
tr:nth-child(even) td { background: var(--bg2) !important; }
blockquote {
  border-left: 4px solid var(--gold) !important;
  background: #1a2a1a !important;
  padding: 10px 16px !important;
  border-radius: 0 8px 8px 0 !important;
  color: var(--text) !important;
}
</style>
<div style="background:linear-gradient(135deg,#1a2332 0%,#0d1b2a 100%);border-radius:16px;padding:36px;margin-bottom:12px;">
  <div style="display:flex;align-items:center;gap:18px;margin-bottom:16px;">
    <div style="background:#1b91ff;color:#fff;font-size:36px;font-weight:900;border-radius:12px;padding:8px 20px;">8</div>
    <div>
      <div style="color:#8ab4d4;font-size:12px;font-weight:600;letter-spacing:3px;text-transform:uppercase;">ЕГЭ по информатике</div>
      <div style="color:#ffffff;font-size:24px;font-weight:800;">Комбинаторика</div>
      
    </div>
  </div>
  <div style="display:flex;gap:8px;flex-wrap:wrap;"><span style="background:#1b3a5c;color:#7ec8e3;font-size:12px;padding:4px 14px;border-radius:20px;font-weight:600;">Python</span> <span style="background:#1b3a5c;color:#7ec8e3;font-size:12px;padding:4px 14px;border-radius:20px;font-weight:600;">Перебор</span> <span style="background:#1b3a5c;color:#7ec8e3;font-size:12px;padding:4px 14px;border-radius:20px;font-weight:600;">Комбинации</span> </div>
</div>

<div style="background:#0a1628;border:1px solid #1b3a5c;border-radius:8px;padding:10px 18px;margin-bottom:16px;display:inline-block;">
<a href="../../README.md" style="color:#1b91ff;text-decoration:none;font-weight:600;">← Назад ко всем заданиям</a>
</div>

# Лекция 8: Комбинаторика в задачах ЕГЭ

## 📋 Обзор типов заданий

В 8 задании проверяется умение правильно собирать комбинации по условию задачи. Всего можно выделить **2 основных типа** задач:

### 1. Задачи с цифрами
**Пример:** Сколько существует десятичных четырёхзначных чисел, в которых все цифры различны и никакие две чётные или две нечётные цифры не стоят рядом?

### 2. Задачи с буквами  
**Пример:** Все шестибуквенные слова, в составе которых могут быть только буквы П, О, Б, Е, Д, А, записаны в алфавитном порядке и пронумерованы начиная с 1.

**Начало списка:**
1. АААААА
2. АААААБ
3. АААААД
4. АААААЕ
5. АААААО
6. АААААП
...

**Задача:** Определите последний чётный номер слова, которое начинается с буквы О и в котором каждая буква встречается ровно один раз.

!> **Примечание:** Слово - последовательность идущих подряд букв, не обязательно осмысленная.

---

## 🔍 Разбор задачи с цифрами (Основная волна 2022)

**Условие:** Определите количество пятизначных чисел, записанных в семеричной системе счисления, которые:
1. Начинаются с четных цифр
2. Не оканчиваются на цифры, меньшие 3  
3. Содержат не более одной цифры 4

### 1. Ручной способ решения

!> **ШАГ 1:** Анализируем условие до первой запятой - "пятизначных чисел". Значит, нам нужно найти количество комбинаций из ровно 5 знаков. Представим их в виде 5 позиций:

![5 позиций](images/i1.jpg)

!> **ШАГ 2:** "В семеричной системе счисления" - система содержит 7 цифр: **0, 1, 2, 3, 4, 5, 6**

!> **ВАЖНО:** Название системы показывает количество цифр в ней, а не максимальную цифру!
!> **ВАЖНО:** Во всех системах счисления есть цифра 0!

![Семеричная система](images/i2.png)

!> **ШАГ 3:** "Начинаются с четных цифр" - четные цифры: 0, 2, 4, 6
!> **КРИТИЧЕСКИ ВАЖНО:** Число не может начинаться с 0! Значит, возможные цифры: **2, 4, 6** (3 варианта)

![Первая позиция](images/i3.png)

!> **ШАГ 4:** "Не оканчиваются на цифры, меньшие 3" - нельзя: 0, 1, 2. Можно: **3, 4, 5, 6** (4 варианта)

![Последняя позиция](images/i4.png)

!> **ШАГ 5:** "Содержат не более одной цифры 4" - либо 0 четверок, либо 1 четверка

#### Случай 1: Без цифры 4 вообще
Цифра 4 убирается из всех позиций. В первой позиции остаются цифры **2,6** (2 варианта), в последней - **3,5,6** (3 варианта), в средних - **0,1,2,3,5,6** (6 вариантов)

![Без четверок](images/i5.png)

In [16]:
x1 = 2 * 6 * 6 * 6 * 3
print(x1)

1296


## Разбор случая с ровно одной цифрой 4

!> **Почему именно 5 видов?** По условию, цифра 4 должна встретиться в комбинации ровно **один раз**. Это означает, что она может располагаться в любой одной из пяти доступных позиций, но только строго в одной. Следовательно, существует ровно **5 различных сценариев** (или «видов» комбинаций), которые мы должны рассмотреть отдельно.

Перейдем к рассмотрению первого сценария.

### Вид 1: Цифра 4 находится на первой позиции
![Цифра 4 на первой позиции](images/i7.png)

!> **Важный вопрос:** Почему для первой позиции теперь возможен только **1 вариант** (сама цифра 4), а не 3, как в общем случае?

**Объяснение:** В первоначальном расчете (без четверки) мы учитывали все четные цифры, кроме 0: `2, 4, 6` (3 варианта). Однако сейчас мы рассматриваем не все возможные четные цифры, а **конкретный сценарий**, где на первом месте гарантированно должна стоять цифра **4**. Если бы мы оставили 3 варианта, это означало бы, что мы учитываем и числа, начинающиеся с 2 или 6, что противоречит условию данного частного случая. Нам нужна **100%-ная уверенность**, что первая цифра — это 4, поэтому выбираем ровно **один конкретный вариант**.

Рассчитаем количество комбинаций для этого случая путем перемножения количества вариантов для каждой позиции:

In [19]:
x2 = 1*6*6*6*3
print(x2)

648


**Позиция 2:**
![5 знаков](images/i8.png)


In [20]:
x3 = 2*1*6*6*3
print(x3)

216


**Позиция 3:**
![5 знаков](images/i9.png)

In [21]:
x4 = 2*6*1*6*3
print(x4)

216


**Позиция 4:**
![5 знаков](images/i10.png)

In [22]:
x5 = 2*6*6*1*3
print(x5)

216


**Позиция 5:**
![5 знаков](images/i11.png)

In [23]:
x6 = 2*6*6*6*1
print(x6)

432


## 📊 Суммирование результатов

Итак, мы провели детальный расчет для каждого из шести возможных сценариев:

1.  **Сценарий 0:** Четвёрток нет вообще → `x1 = 1296`
2.  **Сценарий 1:** Четвёрка на 1-й позиции → `x2 = 648`
3.  **Сценарий 2:** Четвёрка на 2-й позиции → `x3 = 216`
4.  **Сценарий 3:** Четвёрка на 3-й позиции → `x4 = 216`
5.  **Сценарий 4:** Четвёрка на 4-й позиции → `x5 = 216`
6.  **Сценарий 5:** Четвёрка на 5-й позиции → `x6 = 432`

!> **ВАЖНО:** Эти сценарии **взаимоисключают** друг друга (четвёрка не может одновременно стоять в двух разных позициях), поэтому общее количество искомых комбинаций находится простым **суммированием** результатов всех сценариев.

In [25]:
print(x1+x2+x3+x4+x5+x6)

3024


## 2. Решение с помощью Python (Вложенные циклы)

Теперь рассмотрим программный подход к решению той же задачи с использованием вложенных циклов. Данный метод реализует **полный перебор** всех возможных комбинаций с последующей проверкой условий.

### Код с подробными комментариями:

In [56]:
# Создаем счетчик для подсчета подходящих чисел
count = 0

# Перебираем все пятизначные числа в семеричной системе
# Первый цикл - первая цифра числа (d1)
# '246' - четные цифры в семеричной системе (2,4,6)
# Цифра 0 исключена, так как пятизначное число не может начинаться с 0
for d1 in '246':  # первая цифра - четная (2,4,6), но не 0
    # Второй цикл - вторая цифра числа (d2)
    # '0123456' - все возможные цифры семеричной системы
    for d2 in '0123456':
        # Третий цикл - третья цифра числа (d3)
        for d3 in '0123456':
            # Четвертый цикл - четвертая цифра числа (d4)
            for d4 in '0123456':
                # Пятый цикл - пятая цифра числа (d5)
                # '3456' - цифры не меньшие 3 (3,4,5,6)
                # Цифры 0,1,2 исключены по условию
                for d5 in '3456':  # последняя цифра не меньше 3 (3,4,5,6)
                    # Собираем все цифры в одно число (строку)
                    number = d1 + d2 + d3 + d4 + d5
                    
                    # Проверяем количество цифр 4 в числе
                    # Метод count('4') возвращает количество вхождений цифры '4'
                    count_4 = number.count('4')
                    
                    # Проверяем условие: не более одной цифры 4 (0 или 1)
                    if count_4 <= 1:  # не более одной цифры 4
                        # Если условие выполняется, увеличиваем счетчик
                        count += 1

# Выводим общее количество подходящих чисел
print(count)

3024


## Версия кода без комментариев

In [32]:
count = 0

for d1 in '246':
    for d2 in '0123456':
        for d3 in '0123456':
            for d4 in '0123456':
                for d5 in '3456':
                    number = d1 + d2 + d3 + d4 + d5
                    count_4 = number.count('4')
                    if count_4 <= 1:
                        count += 1
print(count)

3024


## 3. Решение с помощью `itertools.product` (Самый оптимальный!)

Этот подход использует мощный инструмент языка Python — модуль `itertools`, который предназначен для создания итераторов для эффективного циклического перебора. Метод `product` позволяет сгенерировать **декартово произведение** входных итераторов, что идеально подходит для нашей задачи перебора всех возможных комбинаций цифр.

### Код с подробными комментариями:

In [36]:
# Импортируем модуль itertools для работы с комбинаторикой
from itertools import *

# Задаем алфавит - все цифры семеричной системы счисления
alp = '0123456'

# Генерируем все возможные комбинации длиной 5 из цифр алфавита
# product() создает декартово произведение, repeat=5 означает 5 позиций
combs = product(alp, repeat=5)

# Инициализируем счетчик для подсчета подходящих чисел
count = 0

# Перебираем все сгенерированные комбинации
for comb in combs:
    # Преобразуем кортеж символов в строку
    # comb - это кортеж вида ('0','1','2','3','4'), join объединяет в "01234"
    comb = ''.join(comb)
    
    # Проверяем первое условие: число начинается с четной цифры
    # В семеричной системе четные цифры: 0,2,4,6
    # Но 0 исключен, так как пятизначное число не может начинаться с 0
    if comb[0] in ['2','4','6']:
        
        # Проверяем второе условие: число не оканчивается на цифры меньшие 3
        # То есть последняя цифра должна быть 3,4,5 или 6
        if comb[-1] in ['3','4','5','6']:
            
            # Проверяем третье условие: содержит не более одной цифры 4
            # Метод count() подсчитывает количество вхождений символа '4'
            if comb.count('4') <= 1:
                # Если все три условия выполнены, увеличиваем счетчик
                count += 1

# Выводим общее количество чисел, удовлетворяющих всем условиям
print(count)

3024


## Версия кода без комментариев

In [35]:
from itertools import *
alp = '0123456'
combs = product(alp,repeat=5)
count = 0
for comb in combs:
    comb = ''.join(comb)
    if comb[0] in ['2','4','6']:
        if comb[-1] in ['3','4','5','6']:
            if comb.count('4') <=1:
                count += 1
print(count)      

3024


## 💡 Итоги и выводы по задаче

Вот и разобрали все **3 способа решения** данной комбинаторной задачи. Каждый из методов имеет свои преимущества:

!> **ВАЖНО:** Для успешной сдачи экзамена **необходимо знать и понимать** как минимум:
1.  **Ручной способ (1)** — для развития логического и комбинаторного мышления, а также для уверенности в результате.
2.  **Хотя бы один программный способ (2 или 3)** — для проверки ручного решения и решения сложных случаев, где легко ошибиться в ручном расчете.

!> **МОЯ РЕКОМЕНДАЦИЯ:** Среди программных способов **решение №3 с `itertools.product`** является наиболее предпочтительным, так как оно:
*   **Более читаемое и лаконичное**
*   **Менее подвержено ошибкам** (меньше строк кода — меньше chance ошибиться)
*   **Легче модифицируется** под изменение условий задачи

---

## 🔍 Важное наблюдение по типам задач

Самые внимательные могли заметить ключевую закономерность:

*   **В задачах с ЧИСЛАМИ** вопрос **почти всегда** стоит об **общем количестве** комбинаций, удовлетворяющих условиям.

*   **В задачах с БУКВАМИ** вопрос может быть гораздо **тоньше и разнообразнее**:
    *   Могут спросить **номер конкретной комбинации** в упорядоченном списке
    *   Задать вопрос про **чётный или нечётный номер** слова
    *   Спросить про **первое** или **последнее** слово с определенным свойством
    *   И другие специфические свойства

Это различие важно помнить при выборе стратегии решения!